projects/storefrontlib/src/cms-components/cart/cart-coupon/cart-coupon.component.ts

```diff
@@ -0,0 +1,87 @@
+import { Component, ElementRef, OnDestroy, OnInit } from '@angular/core';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { Cart, CartService, CartVoucherService } from '@spartacus/core';
+import { Observable } from 'rxjs';
+import { Subscription } from 'rxjs/internal/Subscription';
+import { CartCouponAnchorService } from './cart-coupon-anchor/cart-coupon-anchor.service';
+
+@Component({
+  selector: 'cx-cart-coupon',
+  templateUrl: './cart-coupon.component.html',
+})
+export class CartCouponComponent implements OnInit, OnDestroy {
+  form: FormGroup;
+  btnEnabled: boolean;
+  cartIsLoading: boolean;
+  addVoucherIsLoading$: Observable<boolean>;
+  cart$: Observable<Cart>;
+
+  private subscription = new Subscription();
+
+  constructor(
+    private cartService: CartService,
+    private cartVoucherService: CartVoucherService,
+    private formBuilder: FormBuilder,
+    private element: ElementRef,
+    private cartCouponAnchorService: CartCouponAnchorService
+  ) {}
+
+  ngOnInit() {
+    this.cart$ = this.cartService.getActive();
+
+    this.cartIsLoading = !this.cartService.getLoaded();
+
+    this.cartVoucherService.resetAddVoucherProcessingState();
+
+    this.form = this.formBuilder.group({
+      couponCode: ['', [Validators.required]],
+    });
+
+    this.addVoucherIsLoading$ = this.cartVoucherService.getAddVoucherResultLoading();
+
+    this.subscription.add(
+      this.cartVoucherService
+        .getAddVoucherResultSuccess()
+        .subscribe(success => {
+          this.onSuccess(success);
+        })
+    );
+
+    this.subscription
+      .add(
+        this.cartCouponAnchorService.getEventEmit().subscribe(() => {
+          this.scrollToView();
+        })
+      )
+      .add(
+        this.form.valueChanges.subscribe(
```

Comment:
can we expose this as an observable instead of subscribing to it? It's used in a component template only, so we could use `async` there

Reply:
Done